roachtest: sqlsmith/setup=tpch-sf1/setting=no-mutations failed

[(roachtest).sqlsmith/setup=tpch-sf1/setting=no-mutations failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1703406&tab=buildLog) on [master@c76ad970e73e606bc55a372d93f9d2d6acb32c9c](https://github.com/cockroachdb/cockroach/commits/c76ad970e73e606bc55a372d93f9d2d6acb32c9c):

```
		WITH
			with_52 (col_285) AS (SELECT * FROM (VALUES (NULL)) AS tab_135 (col_285)),
			with_53 (col_286) AS (SELECT * FROM (VALUES (NULL)) AS tab_136 (col_286))
		SELECT
			fnv64(experimental_uuid_v4()::BYTES::BYTES)::INT8 AS col_287,
			e'-D\x1d]4mM\x07':::STRING AS col_288,
			tab_137.c_nationkey AS col_289,
			tab_137.c_acctbal AS col_290,
			(
				WITH
					with_54 (col_291, col_292, col_293)
						AS (
							SELECT
								*
							FROM
								(VALUES (e'JKm_e\x13c#]':::STRING:::NAME, 109498869:::OID, (-1.4834214448928833):::FLOAT8))
									AS tab_138 (col_291, col_292, col_293)
						),
					with_57 (col_297)
						AS (
							SELECT
								*
							FROM
								(
									VALUES
										('1970-01-01 19:01:53.000499+00:00':::TIMESTAMP),
										('1970-01-08 14:32:49.000734+00:00':::TIMESTAMP),
										('1970-01-06 10:35:51.000008+00:00':::TIMESTAMP),
										(
											('1970-01-11 05:16:29.000191+00:00':::TIMESTAMP::TIMESTAMP - (WITH with_55 (col_294) AS (SELECT * FROM (VALUES (e'wQq\x18m':::STRING), (e'D$IEr\f':::STRING)) AS tab_139 (col_294)), with_56 (col_295) AS (SELECT * FROM (VALUES (NULL), ('-59 years -10 mons -590 days -23:55:34.905713':::INTERVAL), ('-81 years -9 mons -853 days -18:36:41.010748':::INTERVAL), ('-49 years -11 mons -886 days -21:17:06.671991':::INTERVAL), ('38 years 573 days 11:05:47.878939':::INTERVAL)) AS tab_140 (col_295)) SELECT NULL AS col_296 FROM defaultdb.public.orders@o_ck AS tab_141 WHERE true ORDER BY tab_141.o_shippriority LIMIT 1:::INT8)::INTERVAL)::TIMESTAMP
										),
										(
											('-62 years -4 mons -295 days -11:01:26.695009':::INTERVAL::INTERVAL + '1970-01-05 01:46:15.000284+00:00':::TIMESTAMP::TIMESTAMP)::TIMESTAMP
										)
								)
									AS tab_142 (col_297)
						),
					with_58 (col_298, col_299)
						AS (
							SELECT
								*
							FROM
								(
									VALUES
										(1589839854:::OID, '15:03:30.191006+08:11:00':::TIMETZ),
										(184626351:::OID, '03:49:07.024113+02:36:00':::TIMETZ),
										(4205316319:::OID, '02:19:25.265741+04:23:00':::TIMETZ),
										(4194327752:::OID, '15:10:31.906008+01:42:00':::TIMETZ),
										(3308814731:::OID, NULL)
								)
									AS tab_143 (col_298, col_299)
						)
				SELECT
					tab_137.c_address AS col_300
				FROM
					with_57 AS cte_ref_17, with_57 AS cte_ref_18
				WHERE
					current_date()::DATE::DATE IN ('1973-11-22':::DATE,)
				ORDER BY
					cte_ref_18.col_297 ASC, cte_ref_18.col_297 ASC, cte_ref_18.col_297
				LIMIT
					1:::INT8
			)
				AS col_301,
			NULL AS col_302,
			e'\U00002603':::STRING AS col_303,
			tab_137.c_phone AS col_304
		FROM
			defaultdb.public.customer@[0] AS tab_137
		LIMIT
			71:::INT8;
```

<details><summary>Repro</summary><p>


Artifacts: [/sqlsmith/setup=tpch-sf1/setting=no-mutations](https://teamcity.cockroachdb.com/viewLog.html?buildId=1703406&tab=artifacts#/sqlsmith/setup=tpch-sf1/setting=no-mutations)
[roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Asqlsmith%2Fsetup%3Dtpch-sf1%2Fsetting%3Dno-mutations.%2A&sort=title&restgroup=false&display=lastcommented+project)

<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(roachtest).sqlsmith/setup=tpch-sf1/setting=no-mutations failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1705428&tab=buildLog) on [master@545c96573d36519c27115278187fdd441f54a64d](https://github.com/cockroachdb/cockroach/commits/545c96573d36519c27115278187fdd441f54a64d):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/20200123-1705428/sqlsmith/setup=tpch-sf1/setting=no-mutations/run_1
	sqlsmith.go:175,sqlsmith.go:198,test_runner.go:719: error: pq: communication error: rpc error: code = Canceled desc = context canceled
		stmt:
		SELECT
			e'\x00':::STRING AS col_497, sum(tab_249.ps_supplycost::FLOAT8)::FLOAT8 AS col_498
		FROM
			defaultdb.public.partsupp@ps_sk AS tab_249
		GROUP BY
			tab_249.ps_comment, tab_249.ps_supplycost
		ORDER BY
			tab_249.ps_supplycost DESC, tab_249.ps_comment, tab_249.ps_supplycost ASC, tab_249.ps_comment ASC
		LIMIT
			40:::INT8;

```

<details><summary>More</summary><p>


Artifacts: [/sqlsmith/setup=tpch-sf1/setting=no-mutations](https://teamcity.cockroachdb.com/viewLog.html?buildId=1705428&tab=artifacts#/sqlsmith/setup=tpch-sf1/setting=no-mutations)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Asqlsmith%2Fsetup%3Dtpch-sf1%2Fsetting%3Dno-mutations.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(roachtest).sqlsmith/setup=tpch-sf1/setting=no-mutations failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1710860&tab=buildLog) on [master@72ce65798f55af352c6a38c306542f63dfba465f](https://github.com/cockroachdb/cockroach/commits/72ce65798f55af352c6a38c306542f63dfba465f):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/20200125-1710860/sqlsmith/setup=tpch-sf1/setting=no-mutations/run_1
	sqlsmith.go:170,sqlsmith.go:200,test_runner.go:719: error: pq: internal error: crdb_internal.get_zone_config(): PrivilegedAccessor not set
		stmt:
		WITH
			with_110 (col_487) AS (SELECT * FROM (VALUES ('\xde':::BYTES)) AS tab_246 (col_487)),
			with_111 (col_488)
				AS (
					SELECT
						*
					FROM
						(
							VALUES
								(e'R6aE\x16':::STRING),
								('':::STRING),
								(e'Me\x1d':::STRING),
								(e'\x1c\x05J\x11':::STRING),
								(NULL),
								('':::STRING)
						)
							AS tab_247 (col_488)
				)
		SELECT
			tab_249.ps_comment AS col_489,
			2137770559400709386:::INT8 AS col_490,
			crdb_internal.get_zone_config(tab_249.ps_availqty::INT8)::BYTES AS col_491
		FROM
			with_110 AS cte_ref_29, defaultdb.public.region@[0] AS tab_248, defaultdb.public.partsupp@primary AS tab_249
		WHERE
			true
		LIMIT
			46:::INT8;

```

<details><summary>More</summary><p>


Artifacts: [/sqlsmith/setup=tpch-sf1/setting=no-mutations](https://teamcity.cockroachdb.com/viewLog.html?buildId=1710860&tab=artifacts#/sqlsmith/setup=tpch-sf1/setting=no-mutations)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Asqlsmith%2Fsetup%3Dtpch-sf1%2Fsetting%3Dno-mutations.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(roachtest).sqlsmith/setup=tpch-sf1/setting=no-mutations failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1793930&tab=buildLog) on [master@752dea867f3aeb142e98c22f8d320ce19041aa8d](https://github.com/cockroachdb/cockroach/commits/752dea867f3aeb142e98c22f8d320ce19041aa8d):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/20200307-1793930/sqlsmith/setup=tpch-sf1/setting=no-mutations/run_1
	sqlsmith.go:169,sqlsmith.go:199,test_runner.go:741: error: pq: internal error: unexpected error from the vectorized runtime: recursive partitioning didn't sufficiently decrease the size of the partition: before 96 MB, after 96 MB
		stmt:
		SELECT
			tab_38.l_receiptdate AS col_68,
			8910261727269596638:::INT8 AS col_69,
			tab_39.s_address AS col_70,
			tab_39.s_acctbal AS col_71,
			tab_39.s_address AS col_72
		FROM
			defaultdb.public.lineitem@l_rd AS tab_38, defaultdb.public.supplier@[0] AS tab_39
		ORDER BY
			tab_39.s_suppkey, tab_38.l_discount DESC
		LIMIT
			48:::INT8;

```

<details><summary>More</summary><p>


Artifacts: [/sqlsmith/setup=tpch-sf1/setting=no-mutations](https://teamcity.cockroachdb.com/viewLog.html?buildId=1793930&tab=artifacts#/sqlsmith/setup=tpch-sf1/setting=no-mutations)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Asqlsmith%2Fsetup%3Dtpch-sf1%2Fsetting%3Dno-mutations.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/sql/colexec/disk_spiller.go
+++ b/pkg/sql/colexec/disk_spiller.go
@@ -167,8 +167,10 @@ func newTwoInputDiskSpiller(
 	return &diskSpillerBase{
 		inputs:                 []Operator{inputOne, inputTwo},
 		inMemoryOp:             inMemoryOp,
+		inMemoryOpInitStatus:   OperatorNotInitialized,
 		inMemoryMemMonitorName: inMemoryMemMonitorName,
 		diskBackedOp:           diskBackedOpConstructor(diskBackedOpInputOne, diskBackedOpInputTwo),
+		distBackedOpInitStatus: OperatorNotInitialized,
 		spillingCallbackFn:     spillingCallbackFn,
 	}
 }
@@ -178,27 +180,30 @@ func newTwoInputDiskSpiller(
 type diskSpillerBase struct {
 	NonExplainable
 
-	inputs      []Operator
-	initialized bool
-	spilled     bool
+	inputs  []Operator
+	spilled bool
 
 	inMemoryOp             bufferingInMemoryOperator
+	inMemoryOpInitStatus   OperatorInitStatus
 	inMemoryMemMonitorName string
 	diskBackedOp           Operator
+	distBackedOpInitStatus OperatorInitStatus
 	spillingCallbackFn     func()
 }
 
+var _ resettableOperator = &diskSpillerBase{}
+
 func (d *diskSpillerBase) Init() {
-	if d.initialized {
+	if d.inMemoryOpInitStatus == OperatorInitialized {
 		return
 	}
-	d.initialized = true
 	// It is possible that Init() call below will hit an out of memory error,
 	// but we decide to bail on this query, so we do not catch internal panics.
 	//
 	// Also note that d.input is the input to d.inMemoryOp, so calling Init()
 	// only on the latter is sufficient.
 	d.inMemoryOp.Init()
+	d.inMemoryOpInitStatus = OperatorInitialized
 }
 
 func (d *diskSpillerBase) Next(ctx context.Context) coldata.Batch {
@@ -218,13 +223,33 @@ func (d *diskSpillerBase) Next(ctx context.Context) coldata.Batch {
 				d.spillingCallbackFn()
 			}
 			d.diskBackedOp.Init()
-			return d.Next(ctx)
+			d.distBackedOpInitStatus = OperatorInitialized
+			return d.diskBackedOp.Next(ctx)
 		}
 		// Either not an out of memory error or an OOM error coming from a
 		// different operator, so we propagate it further.
 		execerror.VectorizedInternalPanic(err)
 	}
 	return batch
+}
+
+func (d *diskSpillerBase) reset() {
+	for _, input := range d.inputs {
+		if r, ok := input.(resetter); ok {
+			r.reset()
+		}
+	}
+	if d.inMemoryOpInitStatus == OperatorInitialized {
+		if r, ok := d.inMemoryOp.(resetter); ok {
+			r.reset()
+		}
+	}
+	if d.distBackedOpInitStatus == OperatorInitialized {
+		if r, ok := d.diskBackedOp.(resetter); ok {
+			r.reset()
+		}
+	}
+	d.spilled = false
 }
 
 func (d *diskSpillerBase) Close() error {
@@ -282,7 +307,7 @@ type bufferExportingOperator struct {
 	firstSourceDone bool
 }
 
-var _ Operator = &bufferExportingOperator{}
+var _ resettableOperator = &bufferExportingOperator{}
 
 func newBufferExportingOperator(
 	firstSource bufferingInMemoryOperator, secondSource Operator,
@@ -305,7 +330,17 @@ func (b *bufferExportingOperator) Next(ctx context.Context) coldata.Batch {
 	batch := b.firstSource.ExportBuffered(b.secondSource)
 	if batch.Length() == 0 {
 		b.firstSourceDone = true
-		return b.Next(ctx)
+		return b.secondSource.Next(ctx)
 	}
 	return batch
+}
+
+func (b *bufferExportingOperator) reset() {
+	if r, ok := b.firstSource.(resetter); ok {
+		r.reset()
+	}
+	if r, ok := b.secondSource.(resetter); ok {
+		r.reset()
+	}
+	b.firstSourceDone = false
 }

--- a/pkg/sql/colexec/execplan.go
+++ b/pkg/sql/colexec/execplan.go
@@ -273,6 +273,9 @@ func isSupported(spec *execinfrapb.ProcessorSpec) (bool, error) {
 // according to ordering.
 // - matchLen specifies the length of the prefix of ordering columns the input
 // is already ordered on.
+// - maxNumberPartitions (when non-zero) overrides the semi-dynamically
+// computed maximum number of partitions that the external sorter will have
+// at once.
 // - processorID is the ProcessorID of the processor core that requested
 // creation of this operator. It is used only to distinguish memory monitors.
 // - post describes the post-processing spec of the processor. It will be used
@@ -286,6 +289,7 @@ func (r *NewColOperatorResult) createDiskBackedSort(
 	inputTypes []coltypes.T,
 	ordering execinfrapb.Ordering,
 	matchLen uint32,
+	maxNumberPartitions int,
 	processorID int32,
 	post *execinfrapb.PostProcessSpec,
 	memMonitorNamePrefix string,
@@ -381,13 +385,16 @@ func (r *NewColOperatorResult) createDiskBackedSort(
 			diskQueueCfg := args.DiskQueueCfg
 			diskQueueCfg.CacheMode = colcontainer.DiskQueueCacheModeReuseCache
 			diskQueueCfg.SetDefaultBufferSizeBytesForCacheMode()
+			if args.TestingKnobs.NumForcedRepartitions != 0 {
+				maxNumberPartitions = args.TestingKnobs.NumForcedRepartitions
+			}
 			return newExternalSorter(
 				ctx,
 				unlimitedAllocator,
 				standaloneMemAccount,
 				input, inputTypes, ordering,
 				execinfra.GetWorkMemLimit(flowCtx.Cfg),
-				args.TestingKnobs.NumForcedRepartitions,
+				maxNumberPartitions,
 				diskQueueCfg,
 				args.FDSemaphore,
 			)
@@ -778,6 +785,13 @@ func NewColOperator(
 							execinfra.GetWorkMemLimit(flowCtx.Cfg),
 							diskQueueCfg,
 							args.FDSemaphore,
+							func(input Operator, inputTypes []coltypes.T, orderingCols []execinfrapb.Ordering_Column, maxNumberPartitions int) (Operator, error) {
+								return result.createDiskBackedSort(
+									ctx, flowCtx, args, input, inputTypes,
+									execinfrapb.Ordering{Columns: orderingCols},
+									0 /* matchLen */, maxNumberPartitions, spec.ProcessorID,
+									&execinfrapb.PostProcessSpec{}, monitorNamePrefix+"-")
+							},
 							args.TestingKnobs.NumForcedRepartitions,
 						)
 					},
@@ -836,8 +850,9 @@ func NewColOperator(
 				ctx, result.createBufferingUnlimitedMemAccount(
 					ctx, flowCtx, "merge-joiner",
 				))
-			result.Op, err = NewMergeJoinOp(
-				unlimitedAllocator, execinfra.GetWorkMemLimit(flowCtx.Cfg), args.DiskQueueCfg, args.FDSemaphore,
+			result.Op, err = newMergeJoinOp(
+				unlimitedAllocator, execinfra.GetWorkMemLimit(flowCtx.Cfg),
+				args.DiskQueueCfg, args.FDSemaphore,
 				joinType, inputs[0], inputs[1], leftPhysTypes, rightPhysTypes,
 				core.MergeJoiner.LeftOrdering.Columns, core.MergeJoiner.RightOrdering.Columns,
 			)
@@ -870,7 +885,7 @@ func NewColOperator(
 			ordering := core.Sorter.OutputOrdering
 			matchLen := core.Sorter.OrderingMatchLen
 			result.Op, err = result.createDiskBackedSort(
-				ctx, flowCtx, args, input, inputTypes, ordering, matchLen,
+				ctx, flowCtx, args, input, inputTypes, ordering, matchLen, 0, /* maxNumberPartitions */
 				spec.ProcessorID, post, "", /* memMonitorNamePrefix */
 			)
 			result.ColumnTypes = spec.Input[0].ColumnTypes
@@ -909,8 +924,8 @@ func NewColOperator(
 						func(input Operator, inputTypes []coltypes.T, orderingCols []execinfrapb.Ordering_Column) (Operator, error) {
 							return result.createDiskBackedSort(
 								ctx, flowCtx, args, input, inputTypes,
-								execinfrapb.Ordering{Columns: orderingCols},
-								0 /* matchLen */, spec.ProcessorID,
+								execinfrapb.Ordering{Columns: orderingCols}, 0, /* matchLen */
+								0 /* maxNumberPartitions */, spec.ProcessorID,
 								&execinfrapb.PostProcessSpec{}, memMonitorsPrefix)
 						},
 					)
@@ -920,7 +935,8 @@ func NewColOperator(
 				} else {
 					if len(wf.Ordering.Columns) > 0 {
 						input, err = result.createDiskBackedSort(
-							ctx, flowCtx, args, input, typs, wf.Ordering, 0, /* matchLen */
+							ctx, flowCtx, args, input, typs,
+							wf.Ordering, 0 /* matchLen */, 0, /* maxNumberPartitions */
 							spec.ProcessorID, &execinfrapb.PostProcessSpec{}, memMonitorsPrefix,
 						)
 					}

--- a/pkg/sql/colexec/external_hash_joiner.go
+++ b/pkg/sql/colexec/external_hash_joiner.go
@@ -13,14 +13,16 @@ package colexec
 import (
 	"context"
 	"fmt"
+	"io"
 	"math"
 
 	"github.com/cockroachdb/cockroach/pkg/col/coldata"
+	"github.com/cockroachdb/cockroach/pkg/col/coltypes"
 	"github.com/cockroachdb/cockroach/pkg/sql/colcontainer"
 	"github.com/cockroachdb/cockroach/pkg/sql/colexec/execerror"
+	"github.com/cockroachdb/cockroach/pkg/sql/execinfrapb"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/errors"
-	"github.com/dustin/go-humanize"
 	"github.com/marusama/semaphore"
 )
 
@@ -46,16 +48,19 @@ const (
 	// so we fall back to using a combination of sort and merge join to process
 	// such partition. After repartitioning, the operator transitions to
 	// externalHJJoinNewPartition state.
-	// TODO(yuzefovich): implement the fallback to sort + merge join.
 	externalHJRecursivePartitioning
 	// externalHJJoinNewPartition indicates that the operator should choose a
-	// partition index and join the corresponding partitions from both sides. We
-	// will only join the partitions if the right side partition fits into memory
-	// (because in-memory hash joiner will fully buffer the right side but will
-	// process left side in the streaming fashion). If there are no partition
-	// indices that the operator can join, it transitions into
-	// externalHJRecursivePartitioning state. If there are no partition indices
-	// left at all to join, the operator transitions to externalHJFinished state.
+	// partition index and join the corresponding partitions from both sides
+	// using the in-memory hash joiner. We will only join the partitions if the
+	// right side partition fits into memory (because in-memory hash joiner will
+	// fully buffer the right side but will process left side in the streaming
+	// fashion). If there are no partition indices that the operator can join, it
+	// transitions into externalHJRecursivePartitioning state. If there are no
+	// partition indices to join using in-memory hash joiner, but there are
+	// indices to join using sort + merge join strategy, the operator transitions
+	// to externalHJSortMergeNewPartition state. If there are no partition
+	// indices left at all to join, the operator transitions to
+	// externalHJFinished state.
 	externalHJJoinNewPartition
 	// externalHJJoining indicates that the operator is currently joining tuples
 	// from the corresponding partitions from both sides. An in-memory hash join
@@ -64,6 +69,18 @@ const (
 	// partitions has been emitted), the external hash joiner transitions to
 	// externalHJJoinNewPartition state.
 	externalHJJoining
+	// externalHJSortMergeNewPartition indicates that the operator should choose
+	// a partition index to join using sort + merge join strategy. If there are
+	// no partition indices for this strategy left, the operator transitions to
+	// externalHJFinished state.
+	externalHJSortMergeNewPartition
+	// externalHJSortMergeJoining indicates that the operator is currently
+	// joining tuples from the corresponding partitions from both sides using
+	// (disk-backed) sort + merge join strategy. Once the in-memory merge joiner
+	// returns a zero-length batch (indicating that full output for the current
+	// partitions has been emitted), the external hash joiner transitions to
+	// externalHJSortMergeNewPartition state.
+	externalHJSortMergeJoining
 	// externalHJFinished indicates that the external hash joiner has emitted all
 	// tuples already and only zero-length batch will be emitted from now on.
 	externalHJFinished
@@ -73,14 +90,13 @@ const (
 	// externalHJRecursivePartitioningSizeDecreaseThreshold determines by how
 	// much the newly-created partitions in the recursive partitioning stage
 	// should be smaller than the "parent" partition in order to consider the
-	// repartitioning "successful". If this threshold is not met, then we fall
-	// back to sort + merge join (which, in a sense, serves as the base case
-	// for "recursion").
+	// repartitioning "successful". If this threshold is not met, then this newly
+	// created partition will be added to sort + merge join list (which, in a
+	// sense, serves as the base case for "recursion").
 	externalHJRecursivePartitioningSizeDecreaseThreshold = 0.05
 	// externalHJDiskQueuesMemFraction determines the fraction of the available
 	// RAM that is allocated for the in-memory cache of disk queues.
-	externalHJDiskQueuesMemFraction      = 0.5
-	externalHJFallbackToSortMergeJoinMsg = "recursive partitioning didn't sufficiently decrease the size of the partition"
+	externalHJDiskQueuesMemFraction = 0.5
 )
 
 // externalHashJoiner is an operator that performs Grace hash join algorithm
@@ -112,7 +128,9 @@ const (
 //
 // If one of the partitions itself runs out of memory, we can recursively apply
 // this algorithm. The partition will be divided into sub-partitions by a new
-// hash function, spilled to disk, and so on.
+// hash function, spilled to disk, and so on. If repartitioning doesn't reduce
+// size of the partitions sufficiently, then such partitions will be handled
+// using the combination of disk-backed sort and merge join operators.
 type externalHashJoiner struct {
 	twoInputNode
 	NonExplainable
@@ -133,9 +151,17 @@ type externalHashJoiner struct {
 	maxNumberActivePartitions int
 	// numBuckets is the number of buckets that a partition is divided into.
 	numBuckets int
-	// partitionsToJoin is a map from partitionIdx to a utility struct. This map
-	// contains all partition indices that need to be joined.
-	partitionsToJoin map[int]*externalHJPartitionInfo
+	// partitionsToJoinUsingInMemHash is a map from partitionIdx to a utility
+	// struct. This map contains all partition indices that need to be joined
+	// using the in-memory hash joiner. If the partition is too big, it will be
+	// tried to be repartitioned; if during repartitioning the size doesn't
+	// decrease enough, it will be added to partitionsToJoinUsingSortMerge.
+	partitionsToJoinUsingInMemHash map[int]*externalHJPartitionInfo
+	// partitionsToJoinUsingSortMerge contains all partition indices that need to
+	// be joined using sort + merge join strategy. Partition indices will be
+	// added into this map if recursive partitioning doesn't seem to make
+	// progress on partition' size reduction.
+	partitionsToJoinUsingSortMerge []int
 	// partitionIdxOffset stores the first "available" partition index to use.
 	// During the partitioning step, all tuples will go into one of the buckets
 	// in [partitionIdxOffset, partitionIdxOffset + numBuckets) range.
@@ -156,8 +182,11 @@ type externalHashJoiner struct {
 	}
 
 	// Join phase variables.
-	leftInMemHashJoinerInput, rightInMemHashJoinerInput *partitionerToOperator
-	inMemHashJoiner                                     *hashJoiner
+	leftJoinerInput, rightJoinerInput *partitionerToOperator
+	inMemHashJoiner                   *hashJoiner
+	// diskBackedSortMerge is a side chain of disk-backed sorters that feed into
+	// disk-backed merge joiner which the external hash joiner can fall back to.
+	diskBackedSortMerge resettableOperator
 
 	memState struct {
 		// maxRightPartitionSizeToJoin indicates the maximum memory size of a
@@ -173,13 +202,6 @@ type externalHashJoiner struct {
 		// is forced to recursively repartition (even if it is otherwise not
 		// needed) before it proceeds to actual join partitions.
 		numForcedRepartitions int
-		// repartitionForced is a (temporary) knob that indicates to the external
-		// hash joiner that the last repartition was forced artificially. If it is
-		// true, then we skip the check whether there was a reasonable decrease
-		// in size.
-		// TODO(yuzefovich): remove this once we have fallback to sort + merge
-		// join.
-		repartitionForced bool
 	}
 }
 
@@ -205,9 +227,6 @@ const (
 // - numForcedRepartitions is a number of times that the external hash joiner
 // is forced to recursively repartition (even if it is otherwise not needed).
 // This should be non-zero only in tests.
-// - diskQueuesUnlimitedAllocator is a (temporary) unlimited allocator that
-// will be used to create dummy queues and will be removed once we have actual
-// disk-backed queues.
 func newExternalHashJoiner(
 	unlimitedAllocator *Allocator,
 	spec hashJoinerSpec,
@@ -215,17 +234,18 @@ func newExternalHashJoiner(
 	memoryLimit int64,
 	diskQueueCfg colcontainer.DiskQueueCfg,
 	fdSemaphore semaphore.Semaphore,
+	createReusableDiskBackedSorter func(input Operator, inputTypes []coltypes.T, orderingCols []execinfrapb.Ordering_Column, maxNumberPartitions int) (Operator, error),
 	numForcedRepartitions int,
 ) Operator {
 	if diskQueueCfg.CacheMode != colcontainer.DiskQueueCacheModeClearAndReuseCache {
 		execerror.VectorizedInternalPanic(errors.Errorf("external hash joiner instantiated with suboptimal disk queue cache mode: %d", diskQueueCfg.CacheMode))
 	}
 	leftPartitioner := colcontainer.NewPartitionedDiskQueue(spec.left.sourceTypes, diskQueueCfg, fdSemaphore, colcontainer.PartitionerStrategyDefault)
-	leftInMemHashJoinerInput := newPartitionerToOperator(
+	leftJoinerInput := newPartitionerToOperator(
 		unlimitedAllocator, spec.left.sourceTypes, leftPartitioner, 0, /* partitionIdx */
 	)
 	rightPartitioner := colcontainer.NewPartitionedDiskQueue(spec.right.sourceTypes, diskQueueCfg, fdSemaphore, colcontainer.PartitionerStrategyDefault)
-	rightInMemHashJoinerInput := newPartitionerToOperator(
+	rightJoinerInput := newPartitionerToOperator(
 		unlimitedAllocator, spec.right.sourceTypes, rightPartitioner, 0, /* partitionIdx */
 	)
 	diskQueuesTotalMemLimit := int(float64(memoryLimit) * externalHJDiskQueuesMemFraction)
@@ -245,6 +265,39 @@ func newExternalHashJoiner(
 		// We need at least two buckets per side to make progress.
 		maxNumberActivePartitions = 4
 	}
+	makeOrderingCols := func(eqCols []uint32) []execinfrapb.Ordering_Column {
+		res := make([]execinfrapb.Ordering_Column, len(eqCols))
+		for i, colIdx := range eqCols {
+			res[i].ColIdx = colIdx
+		}
+		return res
+	}
+	// We need to allocate 2 FDs for the merge joiner plus 2 FDs for reading
+	// the partitions that we need to join using sort + merge join strategy, and
+	// all others are divided between the two inputs.
+	externalSorterMaxNumberPartitions := (maxNumberActivePartitions - 4) / 2
+	leftOrdering := makeOrderingCols(spec.left.eqCols)
+	leftPartitionSorter, err := createReusableDiskBackedSorter(
+		leftJoinerInput, spec.left.sourceTypes, leftOrdering, externalSorterMaxNumberPartitions,
+	)
+	if err != nil {
+		execerror.VectorizedInternalPanic(err)
+	}
+	rightOrdering := makeOrderingCols(spec.right.eqCols)
+	rightPartitionSorter, err := createReusableDiskBackedSorter(
+		rightJoinerInput, spec.right.sourceTypes, rightOrdering, externalSorterMaxNumberPartitions,
+	)
+	if err != nil {
+		execerror.VectorizedInternalPanic(err)
+	}
+	diskBackedSortMerge, err := newMergeJoinOp(
+		unlimitedAllocator, memoryLimit, diskQueueCfg,
+		fdSemaphore, spec.joinType, leftPartitionSorter, rightPartitionSorter,
+		spec.left.sourceTypes, spec.right.sourceTypes, leftOrdering, rightOrdering,
+	)
+	if err != nil {
+		execerror.VectorizedInternalPanic(err)
+	}
 	ehj := &externalHashJoiner{
 		twoInputNode:              newTwoInputNode(leftInput, rightInput),
 		unlimitedAllocator:        unlimitedAllocator,
@@ -258,13 +311,15 @@ func newExternalHashJoiner(
 		// half for the right side.
 		// TODO(yuzefovich): figure out whether we should care about
 		// hj.numBuckets being a power of two (finalizeHash step is faster if so).
-		numBuckets:                maxNumberActivePartitions / 2,
-		partitionsToJoin:          make(map[int]*externalHJPartitionInfo),
-		leftInMemHashJoinerInput:  leftInMemHashJoinerInput,
-		rightInMemHashJoinerInput: rightInMemHashJoinerInput,
+		numBuckets:                     maxNumberActivePartitions / 2,
+		partitionsToJoinUsingInMemHash: make(map[int]*externalHJPartitionInfo),
+		partitionsToJoinUsingSortMerge: make([]int, 0),
+		leftJoinerInput:                leftJoinerInput,
+		rightJoinerInput:               rightJoinerInput,
 		inMemHashJoiner: newHashJoiner(
-			unlimitedAllocator, spec, leftInMemHashJoinerInput, rightInMemHashJoinerInput,
+			unlimitedAllocator, spec, leftJoinerInput, rightJoinerInput,
 		).(*hashJoiner),
+		diskBackedSortMerge: diskBackedSortMerge,
 	}
 	// To simplify the accounting, we will assume that the in-memory hash
 	// joiner's memory usage is equal to the size of the right partition to be
@@ -348,10 +403,10 @@ func (hj *externalHashJoiner) partitionBatch(
 			if err := partitioner.Enqueue(ctx, partitionIdx, scratchBatch); err != nil {
 				execerror.VectorizedInternalPanic(err)
 			}
-			partitionInfo, ok := hj.partitionsToJoin[partitionIdx]
+			partitionInfo, ok := hj.partitionsToJoinUsingInMemHash[partitionIdx]
 			if !ok {
 				partitionInfo = &externalHJPartitionInfo{}
-				hj.partitionsToJoin[partitionIdx] = partitionInfo
+				hj.partitionsToJoinUsingInMemHash[partitionIdx] = partitionInfo
 			}
 			if side == rightSide {
 				partitionInfo.rightParentMemSize = parentMemSize
@@ -418,7 +473,7 @@ StateChanged:
 			// hj.numBuckets being a power of two (finalizeHash step is faster if so).
 			hj.numBuckets = hj.maxNumberActivePartitions - 1
 			hj.tupleDistributor.resetNumOutputs(hj.numBuckets)
-			for parentPartitionIdx, parentPartitionInfo := range hj.partitionsToJoin {
+			for parentPartitionIdx, parentPartitionInfo := range hj.partitionsToJoinUsingInMemHash {
 				for _, side := range []joinSide{leftSide, rightSide} {
 					batch := hj.recursiveScratch.leftBatch
 					partitioner := hj.leftPartitioner
@@ -444,67 +499,76 @@ StateChanged:
 					}
 					// We're done writing to the newly created partitions.
 					// TODO(yuzefovich): we should not release the descriptors here. The
-					// invariant should be: we're enteringexternalHJRecursivePartitioning,
-					// at that stage we have at most numBuckets*2 file descriptors open. At
-					// the top of the state transition, close all open write file
-					// descriptors, which should reduce the open descriptors to 0. Now we
-					// open the two read partitions for 2 file descriptors and whatever
-					// number of write partitions we want. This'll allow us to remove the
-					// call to CloseAllOpen... in the first state as well.
+					// invariant should be: we're entering
+					// externalHJRecursivePartitioning, at that stage we have at most
+					// numBuckets*2 file descriptors open. At the top of the state
+					// transition, close all open write file descriptors, which should
+					// reduce the open descriptors to 0. Now we open the two read'
+					// partitions for 2 file descriptors and whatever number of write
+					// partitions we want. This'll allow us to remove the call to
+					// CloseAllOpen... in the first state as well.
 					if err := partitioner.CloseAllOpenWriteFileDescriptors(); err != nil {
 						execerror.VectorizedInternalPanic(err)
 					}
 				}
-				if !hj.testingKnobs.repartitionForced {
-					// If the repartition was forced, then it is possible that there is
-					// no reduction in size, and we don't want to error out in such case.
-					for idx := 0; idx < hj.numBuckets; idx++ {
-						if partitionInfo, ok := hj.partitionsToJoin[hj.partitionIdxOffset+idx]; ok {
-							before, after := partitionInfo.rightParentMemSize, partitionInfo.rightMemSize
-							if before > 0 {
-								sizeDecrease := 1.0 - float64(after)/float64(before)
-								if sizeDecrease < externalHJRecursivePartitioningSizeDecreaseThreshold {
-									// TODO(yuzefovich): support this case by falling back to
-									// sort + merge join.
-									execerror.VectorizedInternalPanic(errors.Errorf(
-										"%s: before %s, after %s", externalHJFallbackToSortMergeJoinMsg,
-										humanize.Bytes(uint64(before)), humanize.Bytes(uint64(after))))
-								}
+				for idx := 0; idx < hj.numBuckets; idx++ {
+					newPartitionIdx := hj.partitionIdxOffset + idx
+					if partitionInfo, ok := hj.partitionsToJoinUsingInMemHash[newPartitionIdx]; ok {
+						before, after := partitionInfo.rightParentMemSize, partitionInfo.rightMemSize
+						if before > 0 {
+							sizeDecrease := 1.0 - float64(after)/float64(before)
+							if sizeDecrease < externalHJRecursivePartitioningSizeDecreaseThreshold {
+								// We will need to join this partition using sort + merge
+								// join strategy.
+								hj.partitionsToJoinUsingSortMerge = append(hj.partitionsToJoinUsingSortMerge, newPartitionIdx)
+								delete(hj.partitionsToJoinUsingInMemHash, newPartitionIdx)
 							}
 						}
 					}
 				}
-				hj.testingKnobs.repartitionForced = false
 				// We have successfully repartitioned the partitions with index
 				// 'parentPartitionIdx' on both sides, so we delete that index from the
 				// map and proceed on joining the newly created partitions.
-				delete(hj.partitionsToJoin, parentPartitionIdx)
+				delete(hj.partitionsToJoinUsingInMemHash, parentPartitionIdx)
 				hj.partitionIdxOffset += hj.numBuckets
 				hj.state = externalHJJoinNewPartition
 				continue StateChanged
 			}
 
 		case externalHJJoinNewPartition:
-			if hj.testingKnobs.numForcedRepartitions > 0 {
+			if hj.testingKnobs.numForcedRepartitions > 0 && len(hj.partitionsToJoinUsingInMemHash) > 0 {
 				hj.testingKnobs.numForcedRepartitions--
-				hj.testingKnobs.repartitionForced = true
 				hj.state = externalHJRecursivePartitioning
 				continue
 			}
 			// Find next partition that we can join without having to recursively
 			// repartition.
-			for partitionIdx, partitionInfo := range hj.partitionsToJoin {
+			for partitionIdx, partitionInfo := range hj.partitionsToJoinUsingInMemHash {
 				if partitionInfo.rightMemSize <= hj.memState.maxRightPartitionSizeToJoin {
 					// Update the inputs to in-memory hash joiner and reset the latter.
-					hj.leftInMemHashJoinerInput.partitionIdx = partitionIdx
-					hj.rightInMemHashJoinerInput.partitionIdx = partitionIdx
+					hj.leftJoinerInput.partitionIdx = partitionIdx
+					hj.rightJoinerInput.partitionIdx = partitionIdx
 					hj.inMemHashJoiner.reset()
+					delete(hj.partitionsToJoinUsingInMemHash, partitionIdx)
 					hj.state = externalHJJoining
-					delete(hj.partitionsToJoin, partitionIdx)
 					continue StateChanged
 				}
 			}
-			if len(hj.partitionsToJoin) == 0 {
+			if len(hj.partitionsToJoinUsingInMemHash) == 0 {
+				// All partitions to join using the hash joiner have been processed.
+				if len(hj.partitionsToJoinUsingSortMerge) > 0 {
+					// But there are still some partitions to join using sort + merge
+					// join strategy.
+					hj.diskBackedSortMerge.Init()
+					if log.V(2) {
+						log.Info(ctx, fmt.Sprintf(
+							"external hash joiner will join %d partitions using sort + merge join",
+							len(hj.partitionsToJoinUsingSortMerge),
+						))
+					}
+					hj.state = externalHJSortMergeNewPartition
+					continue
+				}
 				// All partitions have been processed, so we transition to finished
 				// state.
 				hj.state = externalHJFinished
@@ -513,6 +577,7 @@ StateChanged:
 			// We have partitions that we cannot join without recursively
 			// repartitioning first, so we transition to the corresponding state.
 			hj.state = externalHJRecursivePartitioning
+			continue
 
 		case externalHJJoining:
 			b := hj.inMemHashJoiner.Next(ctx)
@@ -529,6 +594,39 @@ StateChanged:
 				continue
 			}
 			return b
+
+		case externalHJSortMergeNewPartition:
+			if len(hj.partitionsToJoinUsingSortMerge) == 0 {
+				// All partitions have been processed, so we transition to finished
+				// state.
+				hj.state = externalHJFinished
+				continue
+			}
+			partitionIdx := hj.partitionsToJoinUsingSortMerge[0]
+			hj.partitionsToJoinUsingSortMerge = hj.partitionsToJoinUsingSortMerge[1:]
+			// Update the inputs to sort + merge joiner and reset that chain.
+			hj.leftJoinerInput.partitionIdx = partitionIdx
+			hj.rightJoinerInput.partitionIdx = partitionIdx
+			hj.diskBackedSortMerge.reset()
+			hj.state = externalHJSortMergeJoining
+			continue
+
+		case externalHJSortMergeJoining:
+			b := hj.diskBackedSortMerge.Next(ctx)
+			if b.Length() == 0 {
+				// We're done joining these partitions, so we close them and transition
+				// to joining new ones.
+				if err := hj.leftPartitioner.CloseInactiveReadPartitions(); err != nil {
+					execerror.VectorizedInternalPanic(err)
+				}
+				if err := hj.rightPartitioner.CloseInactiveReadPartitions(); err != nil {
+					execerror.VectorizedInternalPanic(err)
+				}
+				hj.state = externalHJSortMergeNewPartition
+				continue
+			}
+			return b
+
 		case externalHJFinished:
 			if err := hj.Close(); err != nil {
 				execerror.VectorizedInternalPanic(err)
@@ -550,6 +648,11 @@ func (hj *externalHashJoiner) Close() error {
 	}
 	if err := hj.rightPartitioner.Close(); err != nil && retErr == nil {
 		retErr = err
+	}
+	if c, ok := hj.diskBackedSortMerge.(io.Closer); ok {
+		if err := c.Close(); err != nil && retErr == nil {
+			retErr = err
+		}
 	}
 	hj.closed = true
 	return retErr

--- a/pkg/sql/colexec/external_hash_joiner_test.go
+++ b/pkg/sql/colexec/external_hash_joiner_test.go
@@ -13,14 +13,12 @@ package colexec
 import (
 	"context"
 	"fmt"
-	"strings"
 	"testing"
 
 	"github.com/cockroachdb/cockroach/pkg/col/coldata"
 	"github.com/cockroachdb/cockroach/pkg/col/coltypes"
 	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
 	"github.com/cockroachdb/cockroach/pkg/sql/colcontainer"
-	"github.com/cockroachdb/cockroach/pkg/sql/colexec/execerror"
 	"github.com/cockroachdb/cockroach/pkg/sql/execinfra"
 	"github.com/cockroachdb/cockroach/pkg/sql/execinfrapb"
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
@@ -31,6 +29,17 @@ import (
 	"github.com/marusama/semaphore"
 	"github.com/stretchr/testify/require"
 )
+
+// externalHJTestMaxNumberPartitions specifies the minimum number of partitions
+// that the maximum limit on the number of partitions for the external hash
+// joiner can be set to. Here is the reasoning (note that we only describe a
+// point in time when the operator will use the most number of files
+// descriptors simultaneously - the operator is supposed to release FDs once it
+// is done with the corresponding partitions): external hash joiner itself
+// needs at least two partitions per side (i.e. we need at least 4), but during
+// the fallback to sort + merge join we will be using 6 (external sorter on
+// each side will use 2 to read its own partitions and 1 more to write out).
+const externalHJTestMaxNumberPartitions = 6
 
 func TestExternalHashJoiner(t *testing.T) {
 	defer leaktest.AfterTest(t)()
@@ -51,8 +60,6 @@ func TestExternalHashJoiner(t *testing.T) {
 		memAccounts []*mon.BoundAccount
 		memMonitors []*mon.BytesMonitor
 	)
-	// External hash joiner needs at least two partitions per side.
-	const maxNumberPartitions = 4
 	// Test the case in which the default memory is used as well as the case in
 	// which the joiner spills to disk.
 	for _, spillForced := range []bool{false, true} {
@@ -85,7 +92,7 @@ func TestExternalHashJoiner(t *testing.T) {
 						tc.skipAllNullsInjection = true
 					}
 					runHashJoinTestCase(t, tc, func(sources []Operator) (Operator, error) {
-						sem := NewTestingSemaphore(maxNumberPartitions)
+						sem := NewTestingSemaphore(externalHJTestMaxNumberPartitions)
 						semsToCheck = append(semsToCheck, sem)
 						spec := createSpecForHashJoiner(tc)
 						hjOp, accounts, monitors, err := createDiskBackedHashJoiner(
@@ -150,12 +157,10 @@ func TestExternalHashJoinerFallbackToSortMergeJoin(t *testing.T) {
 	var spilled bool
 	queueCfg, cleanup := colcontainerutils.NewTestingDiskQueueCfg(t, true /* inMem */)
 	defer cleanup()
-	// External hash joiner needs at least two partitions per side.
-	const maxNumberPartitions = 4
 	hj, accounts, monitors, err := createDiskBackedHashJoiner(
 		ctx, flowCtx, spec, []Operator{leftSource, rightSource},
 		func() { spilled = true }, queueCfg, 0, /* numForcedRepartitions */
-		NewTestingSemaphore(maxNumberPartitions),
+		NewTestingSemaphore(externalHJTestMaxNumberPartitions),
 	)
 	defer func() {
 		for _, memAccount := range accounts {
@@ -167,15 +172,15 @@ func TestExternalHashJoinerFallbackToSortMergeJoin(t *testing.T) {
 	}()
 	require.NoError(t, err)
 	hj.Init()
-	err = execerror.CatchVectorizedRuntimeError(func() {
-		for b := hj.Next(ctx); b.Length() > 0; b = hj.Next(ctx) {
-		}
-	})
+	// We have a full cross-product, so we should get the number of tuples
+	// squared in the output.
+	expectedTuplesCount := nBatches * nBatches * coldata.BatchSize() * coldata.BatchSize()
+	actualTuplesCount := 0
+	for b := hj.Next(ctx); b.Length() > 0; b = hj.Next(ctx) {
+		actualTuplesCount += b.Length()
+	}
 	require.True(t, spilled)
-	// Currently, we don't have the fallback in place, so we expect an error.
-	// TODO(yuzefovich): change this once we have the fallback.
-	require.NotNil(t, err)
-	require.True(t, strings.Contains(err.Error(), externalHJFallbackToSortMergeJoinMsg))
+	require.Equal(t, expectedTuplesCount, actualTuplesCount)
 }
 
 func BenchmarkExternalHashJoiner(b *testing.B) {

--- a/pkg/sql/colexec/external_sort.go
+++ b/pkg/sql/colexec/external_sort.go
@@ -114,6 +114,7 @@ type externalSorter struct {
 	ordering           execinfrapb.Ordering
 	inMemSorter        resettableOperator
 	partitioner        colcontainer.PartitionedQueue
+	partitionerCreator func() colcontainer.PartitionedQueue
 	// numPartitions is the current number of partitions.
 	numPartitions int
 	// firstPartitionIdx is the index of the first partition to merge next.
@@ -123,7 +124,7 @@ type externalSorter struct {
 	emitter Operator
 }
 
-var _ Operator = &externalSorter{}
+var _ resettableOperator = &externalSorter{}
 
 // newExternalSorter returns a disk-backed general sort operator.
 // - ctx is the same context that standaloneMemAccount was created with.
@@ -137,12 +138,8 @@ var _ Operator = &externalSorter{}
 // limit. The budget *must* be standalone because we don't want to double
 // count the memory (the memory under the batches will be accounted for with
 // the unlimitedAllocator).
-// - diskQueuesUnlimitedAllocator is a (temporary) unlimited allocator that
-// will be used to create dummy queues and will be removed once we have actual
-// disk-backed queues.
 // - maxNumberPartitions (when non-zero) overrides the semi-dynamically
-// computed maximum number of partitions to have at once. It should be
-// non-zero only in tests.
+// computed maximum number of partitions to have at once.
 func newExternalSorter(
 	ctx context.Context,
 	unlimitedAllocator *Allocator,
@@ -171,19 +168,21 @@ func newExternalSorter(
 		// it almost all of the available memory (except for a single output batch
 		// that mergers will use).
 		batchMemSize := estimateBatchSizeBytes(inputTypes, coldata.BatchSize())
-		maxNumberPartitions = (int(memoryLimit) - batchMemSize) / (diskQueueCfg.BufferSizeBytes + batchMemSize)
+		maxNumberPartitions = (int(memoryLimit) - batchMemSize) / diskQueueCfg.BufferSizeBytes
 	}
 	// In order to make progress when merging we have to merge at least two
-	// partitions.
-	if maxNumberPartitions < 2 {
-		maxNumberPartitions = 2
+	// partitions into a new third one.
+	if maxNumberPartitions < 3 {
+		maxNumberPartitions = 3
 	}
 	return &externalSorter{
-		OneInputNode:        NewOneInputNode(inMemSorter),
-		unlimitedAllocator:  unlimitedAllocator,
-		memoryLimit:         memoryLimit,
-		inMemSorter:         inMemSorter,
-		partitioner:         colcontainer.NewPartitionedDiskQueue(inputTypes, diskQueueCfg, fdSemaphore, colcontainer.PartitionerStrategyCloseOnNewPartition),
+		OneInputNode:       NewOneInputNode(inMemSorter),
+		unlimitedAllocator: unlimitedAllocator,
+		memoryLimit:        memoryLimit,
+		inMemSorter:        inMemSorter,
+		partitionerCreator: func() colcontainer.PartitionedQueue {
+			return colcontainer.NewPartitionedDiskQueue(inputTypes, diskQueueCfg, fdSemaphore, colcontainer.PartitionerStrategyCloseOnNewPartition)
+		},
 		inputTypes:          inputTypes,
 		ordering:            ordering,
 		maxNumberPartitions: maxNumberPartitions,
@@ -209,6 +208,9 @@ func (s *externalSorter) Next(ctx context.Context) coldata.Batch {
 				continue
 			}
 			newPartitionIdx := s.firstPartitionIdx + s.numPartitions
+			if s.partitioner == nil {
+				s.partitioner = s.partitionerCreator()
+			}
 			if err := s.partitioner.Enqueue(ctx, newPartitionIdx, b); err != nil {
 				execerror.VectorizedInternalPanic(err)
 			}
@@ -222,10 +224,11 @@ func (s *externalSorter) Next(ctx context.Context) coldata.Batch {
 				// sorter (which will reset inputPartitioningOperator).
 				s.inMemSorter.reset()
 				s.numPartitions++
-				if s.numPartitions == s.maxNumberPartitions {
-					// We have reached the maximum number of active partitions, so we
-					// need to merge them and spill the new partition to disk before we
-					// can proceed on consuming the input.
+				if s.numPartitions == s.maxNumberPartitions-1 {
+					// We have reached the maximum number of active partitions that we
+					// know that we'll be able to merge without exceeding the limit, so
+					// we need to merge all of them and spill the new partition to disk
+					// before we can proceed on consuming the input.
 					s.state = externalSorterRepeatedMerging
 					continue
 				}
@@ -290,12 +293,30 @@ func (s *externalSorter) Next(ctx context.Context) coldata.Batch {
 	}
 }
 
+func (s *externalSorter) reset() {
+	if r, ok := s.input.(resetter); ok {
+		r.reset()
+	}
+	s.state = externalSorterNewPartition
+	if err := s.Close(); err != nil {
+		execerror.VectorizedInternalPanic(err)
+	}
+	s.closed = false
+	s.firstPartitionIdx = 0
+	s.numPartitions = 0
+}
+
 func (s *externalSorter) Close() error {
 	if s.closed {
 		return nil
 	}
+	var err error
+	if s.partitioner != nil {
+		err = s.partitioner.Close()
+		s.partitioner = nil
+	}
 	s.closed = true
-	return s.partitioner.Close()
+	return err
 }
 
 // createMergerForPartitions creates an ordered synchronizer that will merge
@@ -383,5 +404,8 @@ func (o *inputPartitioningOperator) Next(ctx context.Context) coldata.Batch {
 }
 
 func (o *inputPartitioningOperator) reset() {
+	if r, ok := o.input.(resetter); ok {
+		r.reset()
+	}
 	o.standaloneMemAccount.Shrink(o.ctx, o.standaloneMemAccount.Used())
 }

--- a/pkg/sql/colexec/external_sort_test.go
+++ b/pkg/sql/colexec/external_sort_test.go
@@ -52,7 +52,7 @@ func TestExternalSort(t *testing.T) {
 		memAccounts []*mon.BoundAccount
 		memMonitors []*mon.BytesMonitor
 	)
-	const maxNumberPartitions = 2
+	const maxNumberPartitions = 3
 	// Test the case in which the default memory is used as well as the case in
 	// which the joiner spills to disk.
 	for _, spillForced := range []bool{false, true} {
@@ -149,7 +149,7 @@ func TestExternalSortRandomized(t *testing.T) {
 		memAccounts []*mon.BoundAccount
 		memMonitors []*mon.BytesMonitor
 	)
-	const maxNumberPartitions = 2
+	const maxNumberPartitions = 3
 	// Interesting disk spilling scenarios:
 	// 1) The sorter is forced to spill to disk as soon as possible.
 	// 2) The memory limit is dynamically set to repartition twice, this will also

--- a/pkg/sql/colexec/mergejoiner.go
+++ b/pkg/sql/colexec/mergejoiner.go
@@ -12,6 +12,7 @@ package colexec
 
 import (
 	"context"
+	"io"
 	"unsafe"
 
 	"github.com/cockroachdb/cockroach/pkg/col/coldata"
@@ -96,9 +97,25 @@ type mjBuilderCrossProductState struct {
 type mjBufferedGroup struct {
 	*spillingQueue
 	// firstTuple stores a single tuple that was first in the buffered group.
-	firstTuple  []coldata.Vec
-	numTuples   int
-	needToReset bool
+	firstTuple []coldata.Vec
+	numTuples  int
+}
+
+func (bg *mjBufferedGroup) reset() {
+	if err := bg.close(); err != nil {
+		execerror.VectorizedInternalPanic(err)
+	}
+	bg.numTuples = 0
+}
+
+func (bg *mjBufferedGroup) close() error {
+	if bg.spillingQueue != nil {
+		if err := bg.spillingQueue.close(); err != nil {
+			return err
+		}
+		bg.spillingQueue = nil
+	}
+	return nil
 }
 
 // mjProberState contains all the state required to execute in the probing
@@ -115,8 +132,10 @@ type mjProberState struct {
 	// Local buffer for the last left and right groups which is used when the
 	// group ends with a batch and the group on each side needs to be saved to
 	// state in order to be able to continue it in the next batch.
-	lBufferedGroup mjBufferedGroup
-	rBufferedGroup mjBufferedGroup
+	lBufferedGroup            mjBufferedGroup
+	rBufferedGroup            mjBufferedGroup
+	lBufferedGroupNeedToReset bool
+	rBufferedGroupNeedToReset bool
 }
 
 // mjState represents the state of the merge joiner.
@@ -199,11 +218,11 @@ type mergeJoinInput struct {
 // expanding the groups (leftGroups and rightGroups) when a group spans
 // multiple batches.
 
-// NewMergeJoinOp returns a new merge join operator with the given spec that
+// newMergeJoinOp returns a new merge join operator with the given spec that
 // implements sort-merge join. It performs a merge on the left and right input
 // sources, based on the equality columns, assuming both inputs are in sorted
 // order.
-func NewMergeJoinOp(
+func newMergeJoinOp(
 	unlimitedAllocator *Allocator,
 	memoryLimit int64,
 	diskQueueCfg colcontainer.DiskQueueCfg,
@@ -215,7 +234,7 @@ func NewMergeJoinOp(
 	rightTypes []coltypes.T,
 	leftOrdering []execinfrapb.Ordering_Column,
 	rightOrdering []execinfrapb.Ordering_Column,
-) (Operator, error) {
+) (resettableOperator, error) {
 	// TODO(yuzefovich): get rid off "outCols" entirely and plumb the assumption
 	// of outputting all columns into the merge joiner itself.
 	leftOutCols := make([]uint32, len(leftTypes))
@@ -309,6 +328,8 @@ func newMergeJoinBase(
 		rDirections[i] = c.Direction
 	}
 
+	diskQueueCfg.CacheMode = colcontainer.DiskQueueCacheModeReuseCache
+	diskQueueCfg.SetDefaultBufferSizeBytesForCacheMode()
 	base := mergeJoinBase{
 		twoInputNode:       newTwoInputNode(left, right),
 		unlimitedAllocator: unlimitedAllocator,
@@ -387,6 +408,24 @@ type mergeJoinBase struct {
 	}
 }
 
+var _ resetter = &mergeJoinBase{}
+
+func (o *mergeJoinBase) reset() {
+	if r, ok := o.left.source.(resetter); ok {
+		r.reset()
+	}
+	if r, ok := o.right.source.(resetter); ok {
+		r.reset()
+	}
+	o.outputReady = false
+	o.state = mjEntry
+	o.proberState.lBatch = nil
+	o.proberState.rBatch = nil
+	o.proberState.lBufferedGroup.reset()
+	o.proberState.rBufferedGroup.reset()
+	o.resetBuilderCrossProductState()
+}
+
 func (o *mergeJoinBase) getOutColTypes() []coltypes.T {
 	outColTypes := make([]coltypes.T, 0, len(o.left.outCols)+len(o.right.outCols))
 	for _, leftOutCol := range o.left.outCols {
@@ -420,18 +459,10 @@ func (o *mergeJoinBase) initWithOutputBatchSize(outBatchSize int) {
 		o.outputBatchSize = 1<<16 - 1
 	}
 
-	o.proberState.lBufferedGroup.spillingQueue = newSpillingQueue(
-		o.unlimitedAllocator, o.left.sourceTypes, o.memoryLimit,
-		o.diskQueueCfg, o.fdSemaphore, coldata.BatchSize(),
-	)
 	o.proberState.lBufferedGroup.firstTuple = make([]coldata.Vec, len(o.left.sourceTypes))
 	for colIdx, colType := range o.left.sourceTypes {
 		o.proberState.lBufferedGroup.firstTuple[colIdx] = o.unlimitedAllocator.NewMemColumn(colType, 1)
 	}
-	o.proberState.rBufferedGroup.spillingQueue = newRewindableSpillingQueue(
-		o.unlimitedAllocator, o.right.sourceTypes, o.memoryLimit,
-		o.diskQueueCfg, o.fdSemaphore, coldata.BatchSize(),
-	)
 	o.proberState.rBufferedGroup.firstTuple = make([]coldata.Vec, len(o.right.sourceTypes))
 	for colIdx, colType := range o.right.sourceTypes {
 		o.proberState.rBufferedGroup.firstTuple[colIdx] = o.unlimitedAllocator.NewMemColumn(colType, 1)
@@ -472,6 +503,12 @@ func (o *mergeJoinBase) appendToBufferedGroup(
 	if input == &o.left {
 		sourceTypes = o.left.sourceTypes
 		bufferedGroup = &o.proberState.lBufferedGroup
+		if bufferedGroup.spillingQueue == nil {
+			bufferedGroup.spillingQueue = newSpillingQueue(
+				o.unlimitedAllocator, o.left.sourceTypes, o.memoryLimit,
+				o.diskQueueCfg, o.fdSemaphore, coldata.BatchSize(),
+			)
+		}
 		// TODO(yuzefovich): uncomment when spillingQueue actually copies the
 		// enqueued batches when those are kept in memory.
 		//if o.scratch.lBufferedGroupBatch == nil {
@@ -481,6 +518,12 @@ func (o *mergeJoinBase) appendToBufferedGroup(
 	} else {
 		sourceTypes = o.right.sourceTypes
 		bufferedGroup = &o.proberState.rBufferedGroup
+		if bufferedGroup.spillingQueue == nil {
+			bufferedGroup.spillingQueue = newRewindableSpillingQueue(
+				o.unlimitedAllocator, o.right.sourceTypes, o.memoryLimit,
+				o.diskQueueCfg, o.fdSemaphore, coldata.BatchSize(),
+			)
+		}
 		// TODO(yuzefovich): uncomment when spillingQueue actually copies the
 		// enqueued batches when those are kept in memory.
 		//if o.scratch.rBufferedGroupBatch == nil {
@@ -554,19 +597,13 @@ func (o *mergeJoinBase) initProberState(ctx context.Context) {
 		o.proberState.rIdx, o.proberState.rBatch = 0, o.right.source.Next(ctx)
 		o.proberState.rLength = o.proberState.rBatch.Length()
 	}
-	if o.proberState.lBufferedGroup.needToReset {
-		if o.proberState.lBufferedGroup.spillingQueue != nil {
-			o.proberState.lBufferedGroup.reset()
-		}
-		o.proberState.lBufferedGroup.numTuples = 0
-		o.proberState.lBufferedGroup.needToReset = false
+	if o.proberState.lBufferedGroupNeedToReset {
+		o.proberState.lBufferedGroup.reset()
+		o.proberState.lBufferedGroupNeedToReset = false
 	}
-	if o.proberState.rBufferedGroup.needToReset {
-		if o.proberState.rBufferedGroup.spillingQueue != nil {
-			o.proberState.rBufferedGroup.reset()
-		}
-		o.proberState.rBufferedGroup.numTuples = 0
-		o.proberState.rBufferedGroup.needToReset = false
+	if o.proberState.rBufferedGroupNeedToReset {
+		o.proberState.rBufferedGroup.reset()
+		o.proberState.rBufferedGroupNeedToReset = false
 	}
 }
 
@@ -673,4 +710,28 @@ func (o *mergeJoinBase) finishProbe(ctx context.Context) {
 		o.proberState.rBatch,
 		o.proberState.rIdx,
 	)
+}
+
+func (o *mergeJoinBase) Close() error {
+	var lastErr error
+	for _, op := range []Operator{o.left.source, o.right.source} {
+		if c, ok := op.(io.Closer); ok {
+			if err := c.Close(); err != nil {
+				lastErr = err
+			}
+		}
+	}
+	if o.proberState.lBufferedGroup.spillingQueue != nil {
+		if err := o.proberState.lBufferedGroup.close(); err != nil {
+			lastErr = err
+		}
+		o.proberState.lBufferedGroup.spillingQueue = nil
+	}
+	if o.proberState.rBufferedGroup.spillingQueue != nil {
+		if err := o.proberState.rBufferedGroup.close(); err != nil {
+			lastErr = err
+		}
+		o.proberState.rBufferedGroup.spillingQueue = nil
+	}
+	return lastErr
 }

--- a/pkg/sql/colexec/mergejoiner_test.go
+++ b/pkg/sql/colexec/mergejoiner_test.go
@@ -1592,13 +1592,10 @@ func TestFullOuterMergeJoinWithMaximumNumberOfGroups(t *testing.T) {
 				}
 				leftSource := newChunkingBatchSource(typs, colsLeft, nTuples)
 				rightSource := newChunkingBatchSource(typs, colsRight, nTuples)
-				a, err := NewMergeJoinOp(
-					testAllocator, defaultMemoryLimit, queueCfg, NewTestingSemaphore(mjFDLimit),
-					sqlbase.FullOuterJoin,
-					leftSource,
-					rightSource,
-					typs,
-					typs,
+				a, err := newMergeJoinOp(
+					testAllocator, defaultMemoryLimit, queueCfg,
+					NewTestingSemaphore(mjFDLimit), sqlbase.FullOuterJoin,
+					leftSource, rightSource, typs, typs,
 					[]execinfrapb.Ordering_Column{{ColIdx: 0, Direction: execinfrapb.Ordering_Column_ASC}},
 					[]execinfrapb.Ordering_Column{{ColIdx: 0, Direction: execinfrapb.Ordering_Column_ASC}},
 				)
@@ -1669,14 +1666,10 @@ func TestMergeJoinerMultiBatch(t *testing.T) {
 					leftSource := newChunkingBatchSource(typs, cols, nTuples)
 					rightSource := newChunkingBatchSource(typs, cols, nTuples)
 
-					a, err := NewMergeJoinOp(
-						testAllocator,
-						defaultMemoryLimit, queueCfg, NewTestingSemaphore(mjFDLimit),
-						sqlbase.InnerJoin,
-						leftSource,
-						rightSource,
-						typs,
-						typs,
+					a, err := newMergeJoinOp(
+						testAllocator, defaultMemoryLimit,
+						queueCfg, NewTestingSemaphore(mjFDLimit), sqlbase.InnerJoin,
+						leftSource, rightSource, typs, typs,
 						[]execinfrapb.Ordering_Column{{ColIdx: 0, Direction: execinfrapb.Ordering_Column_ASC}},
 						[]execinfrapb.Ordering_Column{{ColIdx: 0, Direction: execinfrapb.Ordering_Column_ASC}},
 					)
@@ -1752,14 +1745,10 @@ func TestMergeJoinerMultiBatchRuns(t *testing.T) {
 					leftSource := newChunkingBatchSource(typs, cols, nTuples)
 					rightSource := newChunkingBatchSource(typs, cols, nTuples)
 
-					a, err := NewMergeJoinOp(
-						testAllocator,
-						defaultMemoryLimit, queueCfg, NewTestingSemaphore(mjFDLimit),
-						sqlbase.InnerJoin,
-						leftSource,
-						rightSource,
-						typs,
-						typs,
+					a, err := newMergeJoinOp(
+						testAllocator, defaultMemoryLimit,
+						queueCfg, NewTestingSemaphore(mjFDLimit), sqlbase.InnerJoin,
+						leftSource, rightSource, typs, typs,
 						[]execinfrapb.Ordering_Column{{ColIdx: 0, Direction: execinfrapb.Ordering_Column_ASC}, {ColIdx: 1, Direction: execinfrapb.Ordering_Column_ASC}},
 						[]execinfrapb.Ordering_Column{{ColIdx: 0, Direction: execinfrapb.Ordering_Column_ASC}, {ColIdx: 1, Direction: execinfrapb.Ordering_Column_ASC}},
 					)
@@ -1885,14 +1874,10 @@ func TestMergeJoinerRandomized(t *testing.T) {
 							leftSource := newChunkingBatchSource(typs, lCols, nTuples)
 							rightSource := newChunkingBatchSource(typs, rCols, nTuples)
 
-							a, err := NewMergeJoinOp(
-								testAllocator,
-								defaultMemoryLimit, queueCfg, NewTestingSemaphore(mjFDLimit),
-								sqlbase.InnerJoin,
-								leftSource,
-								rightSource,
-								typs,
-								typs,
+							a, err := newMergeJoinOp(
+								testAllocator, defaultMemoryLimit,
+								queueCfg, NewTestingSemaphore(mjFDLimit), sqlbase.InnerJoin,
+								leftSource, rightSource, typs, typs,
 								[]execinfrapb.Ordering_Column{{ColIdx: 0, Direction: execinfrapb.Ordering_Column_ASC}},
 								[]execinfrapb.Ordering_Column{{ColIdx: 0, Direction: execinfrapb.Ordering_Column_ASC}},
 							)
@@ -1993,10 +1978,9 @@ func BenchmarkMergeJoiner(b *testing.B) {
 
 				benchMemAccount.Clear(ctx)
 				base, err := newMergeJoinBase(
-					NewAllocator(ctx, &benchMemAccount), defaultMemoryLimit, queueCfg, NewTestingSemaphore(mjFDLimit),
-					leftSource, rightSource,
-					[]uint32{0, 1}, []uint32{2, 3},
-					sourceTypes, sourceTypes,
+					NewAllocator(ctx, &benchMemAccount), defaultMemoryLimit, queueCfg,
+					NewTestingSemaphore(mjFDLimit), leftSource, rightSource,
+					[]uint32{0, 1}, []uint32{2, 3}, sourceTypes, sourceTypes,
 					[]execinfrapb.Ordering_Column{{ColIdx: 0, Direction: execinfrapb.Ordering_Column_ASC}},
 					[]execinfrapb.Ordering_Column{{ColIdx: 0, Direction: execinfrapb.Ordering_Column_ASC}},
 				)
@@ -2025,10 +2009,9 @@ func BenchmarkMergeJoiner(b *testing.B) {
 
 				benchMemAccount.Clear(ctx)
 				base, err := newMergeJoinBase(
-					NewAllocator(ctx, &benchMemAccount), defaultMemoryLimit, queueCfg, NewTestingSemaphore(mjFDLimit),
-					leftSource, rightSource,
-					[]uint32{0, 1}, []uint32{2, 3},
-					sourceTypes, sourceTypes,
+					NewAllocator(ctx, &benchMemAccount), defaultMemoryLimit, queueCfg,
+					NewTestingSemaphore(mjFDLimit), leftSource, rightSource,
+					[]uint32{0, 1}, []uint32{2, 3}, sourceTypes, sourceTypes,
 					[]execinfrapb.Ordering_Column{{ColIdx: 0, Direction: execinfrapb.Ordering_Column_ASC}},
 					[]execinfrapb.Ordering_Column{{ColIdx: 0, Direction: execinfrapb.Ordering_Column_ASC}},
 				)
@@ -2059,10 +2042,9 @@ func BenchmarkMergeJoiner(b *testing.B) {
 
 				benchMemAccount.Clear(ctx)
 				base, err := newMergeJoinBase(
-					NewAllocator(ctx, &benchMemAccount), defaultMemoryLimit, queueCfg, NewTestingSemaphore(mjFDLimit),
-					leftSource, rightSource,
-					[]uint32{0, 1}, []uint32{2, 3},
-					sourceTypes, sourceTypes,
+					NewAllocator(ctx, &benchMemAccount), defaultMemoryLimit, queueCfg,
+					NewTestingSemaphore(mjFDLimit), leftSource, rightSource,
+					[]uint32{0, 1}, []uint32{2, 3}, sourceTypes, sourceTypes,
 					[]execinfrapb.Ordering_Column{{ColIdx: 0, Direction: execinfrapb.Ordering_Column_ASC}},
 					[]execinfrapb.Ordering_Column{{ColIdx: 0, Direction: execinfrapb.Ordering_Column_ASC}},
 				)

--- a/pkg/sql/colexec/mergejoiner_tmpl.go
+++ b/pkg/sql/colexec/mergejoiner_tmpl.go
@@ -1241,8 +1241,8 @@ func (o *mergeJoin_JOIN_TYPE_STRINGOp) setBuilderSourceToBufferedGroup(ctx conte
 	// We cannot yet reset the buffered groups because the builder will be taking
 	// input from them. The actual reset will take place on the next call to
 	// initProberState().
-	o.proberState.lBufferedGroup.needToReset = true
-	o.proberState.rBufferedGroup.needToReset = true
+	o.proberState.lBufferedGroupNeedToReset = true
+	o.proberState.rBufferedGroupNeedToReset = true
 }
 
 // exhaustLeftSource sets up the builder to process any remaining tuples from
@@ -1482,17 +1482,11 @@ func (o *mergeJoin_JOIN_TYPE_STRINGOp) Next(ctx context.Context) coldata.Batch {
 				return o.output
 			}
 		case mjDone:
-			if o.proberState.lBufferedGroup.spillingQueue != nil {
-				if err := o.proberState.lBufferedGroup.close(); err != nil {
-					execerror.VectorizedInternalPanic(err)
-				}
-				o.proberState.lBufferedGroup.spillingQueue = nil
+			if err := o.proberState.lBufferedGroup.close(); err != nil {
+				execerror.VectorizedInternalPanic(err)
 			}
-			if o.proberState.rBufferedGroup.spillingQueue != nil {
-				if err := o.proberState.rBufferedGroup.close(); err != nil {
-					execerror.VectorizedInternalPanic(err)
-				}
-				o.proberState.rBufferedGroup.spillingQueue = nil
+			if err := o.proberState.rBufferedGroup.close(); err != nil {
+				execerror.VectorizedInternalPanic(err)
 			}
 			return coldata.ZeroBatch
 		default:

--- a/pkg/sql/colexec/sort.go
+++ b/pkg/sql/colexec/sort.go
@@ -159,9 +159,6 @@ func (p *allSpooler) getValues(i int) coldata.Vec {
 }
 
 func (p *allSpooler) getNumTuples() int {
-	if !p.spooled {
-		execerror.VectorizedInternalPanic("getNumTuples() is called before spool()")
-	}
 	return p.bufferedTuples.Length()
 }
 
@@ -186,12 +183,12 @@ func (p *allSpooler) getWindowedBatch(startIdx, endIdx int) coldata.Batch {
 }
 
 func (p *allSpooler) reset() {
-	p.spooled = false
-	p.bufferedTuples.SetLength(0)
-	p.bufferedTuples.ResetInternalBatch()
 	if r, ok := p.input.(resetter); ok {
 		r.reset()
 	}
+	p.spooled = false
+	p.bufferedTuples.SetLength(0)
+	p.bufferedTuples.ResetInternalBatch()
 }
 
 type sortOp struct {

--- a/pkg/sql/colexec/stats_test.go
+++ b/pkg/sql/colexec/stats_test.go
@@ -85,9 +85,9 @@ func TestVectorizedStatsCollector(t *testing.T) {
 		rightInput := NewVectorizedStatsCollector(rightSource, 1 /* id */, true /* isStall */, timeutil.NewTestStopWatch(timeSource.Now))
 		rightInput.SetOutputWatch(mjInputWatch)
 
-		mergeJoiner, err := NewMergeJoinOp(
-			testAllocator, defaultMemoryLimit, queueCfg, NewTestingSemaphore(4),
-			sqlbase.InnerJoin, leftInput, rightInput,
+		mergeJoiner, err := newMergeJoinOp(
+			testAllocator, defaultMemoryLimit, queueCfg,
+			NewTestingSemaphore(4), sqlbase.InnerJoin, leftInput, rightInput,
 			[]coltypes.T{coltypes.Int64}, []coltypes.T{coltypes.Int64},
 			[]execinfrapb.Ordering_Column{{ColIdx: 0}},
 			[]execinfrapb.Ordering_Column{{ColIdx: 0}},

